gcc/rust/parse/rust-parse-impl.h

```diff
@@ -12023,6 +12023,11 @@ Parser<ManagedTokenSource>::parse_expr (int right_binding_power,
   if (restrictions.expr_can_be_null)
     {
       TokenId id = current_token->get_id ();
+      if (restrictions.stop_on_token && (id == MATCH_ARROW || id == COMMA))
```

Comment:
you're not setting stop_on_token to true so you're not entering these conditional either.